dimod/include/dimod/quadratic_model.h

```diff
@@ -965,160 +89,57 @@ class BinaryQuadraticModel : public QuadraticModelBase<Bias, Index> {
-        // finally sort and sum the neighborhoods we touched
-        for (size_type i = 0; i < counts.size(); ++i) {
-            if (counts[i] > 0) {
-                base_type::adj_[i].sort_and_sum();
-            }
-        }
+        base_type::add_quadratic(row_iterator, col_iterator, bias_iterator, length);
     }
 
     /// Add one (disconnected) variable to the BQM and return its index.
-    index_type add_variable() {
-        index_type vi = this->num_variables();
-        this->resize(vi + 1);
-        return vi;
-    }
+    index_type add_variable() { return base_type::add_variable(); }
 
     /// Change the vartype of the binary quadratic model
     void change_vartype(Vartype vartype) {
-        if (vartype == vartype_) return;  // nothing to do
-
-        bias_type lin_mp, lin_offset_mp, quad_mp, quad_offset_mp, lin_quad_mp;
-        if (vartype == Vartype::BINARY) {
-            lin_mp = 2;
-            lin_offset_mp = -1;
-            quad_mp = 4;
-            lin_quad_mp = -2;
-            quad_offset_mp = .5;
-        } else if (vartype == Vartype::SPIN) {
-            lin_mp = .5;
-            lin_offset_mp = .5;
-            quad_mp = .25;
-            lin_quad_mp = .25;
-            quad_offset_mp = .125;
-        } else {
-            throw std::logic_error("unexpected vartype");
-        }
-
-        for (size_type ui = 0; ui < base_type::num_variables(); ++ui) {
-            bias_type lbias = base_type::linear_biases_[ui];
-
-            base_type::linear_biases_[ui] *= lin_mp;
-            base_type::offset_ += lin_offset_mp * lbias;
-
-            auto begin = base_type::adj_[ui].begin();
-            auto end = base_type::adj_[ui].end();
-            for (auto nit = begin; nit != end; ++nit) {
-                bias_type qbias = (*nit).second;
-
-                (*nit).second *= quad_mp;
-                base_type::linear_biases_[ui] += lin_quad_mp * qbias;
-                base_type::offset_ += quad_offset_mp * qbias;
-            }
-        }
-
-        vartype_ = vartype;
+        if (vartype == this->vartype_) return;  // nothing to do
+        base_type::change_vartypes(this->vartype_, vartype);
+        this->vartype_ = vartype;
     }
 
-    bias_type lower_bound(index_type v) const {
-        return vartype_info<bias_type>::default_min(this->vartype_);
-    }
+    bias_type lower_bound() const { return vartype_info<bias_type>::min(this->vartype_); }
 
-    /**
-     *  Return the number of interactions in the quadratic model.
-     *
-     * `O(num_variables)` complexity.
-     */
-    size_type num_interactions() const {
-        // we can do better than QuadraticModelBase by not needing to account
-        // for self-loops
-        size_type count = 0;
-        for (auto it = this->adj_.begin(); it != this->adj_.end(); ++it) {
-            count += it->size();
-        }
-        return count / 2;
+    bias_type lower_bound(index_type v) const {
+        return vartype_info<bias_type>::min(this->vartype_);
```

Comment:
What is the use of v as an argument here ? Did you mean to use the variable type of v ?

Reply:
Allows for a uniform API between BQM and QM. I know it's unused.